src/main/java/io/confluent/connect/jdbc/source/MultiColumnTimestampHelper.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright 2017 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.connect.jdbc.source;
+
+import io.confluent.connect.jdbc.util.JdbcUtils;
+import io.confluent.connect.jdbc.util.StringUtils;
+import java.sql.Timestamp;
+import java.util.List;
+import org.apache.kafka.connect.data.Struct;
+
+/**
+ * MultiColumnTimestampHelper handles generation of queries when multiple timestamp
+ * columns are specified. Based on the input order, it selects the first not null
+ * timestamp value to detect new or modified rows.
+ */
+public class MultiColumnTimestampHelper implements TimestampHelper {
+
+  private List<String> timestampColumns;
+
+  public MultiColumnTimestampHelper(List<String> timestampColumns) {
+    this.timestampColumns = timestampColumns;
+  }
+
+  @Override
+  public void addWhereClause(StringBuilder builder,
+      String quoteString, String incrementingColumn) {
+    List<String> quotedList = JdbcUtils.quoteList(timestampColumns, quoteString);
+    String quotedTimestampColumn = StringUtils.join(quotedList, ",");
+    builder.append(" WHERE COALESCE(");
+    builder.append(quotedTimestampColumn);
+    builder.append(") < ? AND ((COALESCE(");
+    builder.append(quotedTimestampColumn);
+    builder.append(") = ? AND ");
+    builder.append(JdbcUtils.quoteString(incrementingColumn, quoteString));
+    builder.append(" > ?");
+    builder.append(") OR COALESCE(");
+    builder.append(quotedTimestampColumn);
+    builder.append(") > ?)");
+    builder.append(" ORDER BY COALESCE(");
+    builder.append(quotedTimestampColumn);
+    builder.append("),");
+    builder.append(JdbcUtils.quoteString(incrementingColumn, quoteString));
+    builder.append(" ASC");
+  }
+
+  @Override
+  public void addWhereClause(StringBuilder builder, String quoteString) {
+    List<String> quotedList = JdbcUtils.quoteList(timestampColumns, quoteString);
+    String quotedTimestampColumn = StringUtils.join(quotedList, ",");
+    builder.append(" WHERE COALESCE(");
+    builder.append(quotedTimestampColumn);
+    builder.append(") > ? AND COALESCE(");
```

Comment:
COALESCE returns the first non null timestamp in this case. Wouldn't we need the largest timestamp for every row? In the (created-date, modified-date) example, COALESCE could return created-date for every query, while the modified-date is the column which is being updated.

Reply:
We have to be able to pull out the timestamp value that satisfied the criteria, and if there are multiple timestamps there's no guarantee we know which one actually did. So the current solution is to have the column names be ordered and to grab the first non-null timestamp value, and this corresponds to the `COALESCE` function.

So, a user whose table contains multiple timestamp columns might need to make sure they order the timestamp columns with the preferred one first.

Reply:
Got it. Should we clear up the documentation to reflect this more clearly?

Reply:
+1. I'll add a suggestion in a line comment.

Reply:
I do feel that @wicknicks has a good point here in that this does make it error prone and easy for someone to make a mistake.

Reply:
@mikkin we need to extract the same timestamp value from the record that satisfied the criteria. If we use `COALESCE` in the criteria, then that will select a row if and only if the first non-null column value is greater than the timestamp we provide when querying. That means that when we extract the value, we have to pull the first non-null timestamp column value from the record.

However, if we were to instead do something a bit different and build the criteria with `(x IS NOT NULL AND x > ?)` ORed together for each timestamp column, then we would have to do the same thing to extract the timestamp from the record: loop through the timestamp column values and look for the first non-null value _that is greater than the timestamp used in the criteria_.

Using `COALESCE` is nice because we still only have a single `?` parameter in our prepared statement, whereas using the `(x IS NOT NULL AND x > ?)` criteria for each timestamp column means we have one `?` parameter _for each timestamp column_, and that will change how we set the values on the prepared statement just before we execute it. 

I think for simplicity we should stick with `COALESCE` because it's easier to understand and explain, and because the implementation is simpler. That means that the current logic is correct, although we still need to improve the documentation to make the behavior clear.

Reply:
Also, `COALESCE` is used in the `ORDER BY` clause to ensure the rows are ordered correctly, so that's another reason to use `COALESCE`.

Reply:
I see. We want the DB to order by timestamps. Fair enough. Any reason why we are not using this `GREATEST` function?  (https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions060.htm)

Reply:
`GREATEST` returns `NULL` if any value is null.

Reply:
You can just protect it with `GREATEST ( COALESCE (date1, 0) )` (https://stackoverflow.com/a/9831901)

Reply:
We should really use `ORDER BY ASC NULLS LAST`, but not DBMSes support that `NULLS LAST|FIRST` clause.

From what I understand, only Oracle and MySQL have `GREATEST` function.